File: filesync/src/main/java/de/mel/filesync/data/conflict/ConflictSolver.java
package de.mel.filesync.data.conflict;

import de.mel.Lok;
import de.mel.auth.file.AbstractFile;
import de.mel.auth.file.IFile;
import de.mel.auth.tools.N;
import de.mel.auth.tools.Order;
import de.mel.filesync.data.FileSyncStrings;
import de.mel.filesync.data.fs.RootDirectory;
import de.mel.filesync.index.InitialIndexConflictHelper;
import de.mel.filesync.service.sync.SyncStageMerger;
import de.mel.filesync.sql.*;
import de.mel.filesync.sql.dao.FsDao;
import de.mel.filesync.sql.dao.StageDao;
import de.mel.sql.SqlQueriesException;

import java.io.File;
import java.util.*;
import java.util.concurrent.Semaphore;

/**
 * Created by xor on 5/30/17.
 */
public class ConflictSolver extends SyncStageMerger {
    private final String identifier;
    private final StageSet serverStageSet, localStageSet;
    private final RootDirectory rootDirectory;
    private final Long basedOnVersion;
    protected Map<String, Conflict> deletedParents = new HashMap<>();
    private Map<String, Conflict> conflicts = new HashMap<>();
    private Order order;
    private Map<Long, Long> oldeNewIdMap, oldeObsoleteMap;
    private StageDao stageDao;
    /**
     * obsoleteStageSet: if you decide to go for the server StageSet, some local (but not yet committed changes) must be deleted
     */
    private StageSet mergeStageSet, obsoleteStageSet;
    private final Order obsoleteOrder;


    private FsDao fsDao;
    private List<ConflictSolverListener> listeners = new ArrayList<>();
    private Semaphore listenerSemaphore = new Semaphore(1, true);
    private boolean obsolete = false;
    private boolean solving = false;
    private String conflictHelperUuid;

    public ConflictSolver(FileSyncDatabaseManager fileSyncDatabaseManager, StageSet serverStageSet, StageSet localStageSet) throws SqlQueriesException {
        super(serverStageSet.getId().v(), localStageSet.getId().v());
        this.rootDirectory = fileSyncDatabaseManager.getFileSyncSettings().getRootDirectory();
        this.serverStageSet = serverStageSet;
        this.localStageSet = localStageSet;
        stageDao = fileSyncDatabaseManager.getStageDao();
        fsDao = fileSyncDatabaseManager.getFsDao();
        identifier = createIdentifier(serverStageSet.getId().v(), localStageSet.getId().v());
        basedOnVersion = localStageSet.getBasedOnVersion().v() >= serverStageSet.getBasedOnVersion().v() ? localStageSet.getBasedOnVersion().v() : serverStageSet.getBasedOnVersion().v();
        obsoleteStageSet = stageDao.createStageSet(FileSyncStrings.STAGESET_SOURCE_SERVER, FileSyncStrings.STAGESET_STATUS_DELETE, serverStageSet.getOriginCertId().v(), serverStageSet.getOriginServiceUuid().v(), serverStageSet.getVersion().v(), basedOnVersion);
        obsoleteOrder = new Order();
    }

    public StageSet getServerStageSet() {
        return serverStageSet;
    }

    public StageSet getLocalStageSet() {
        return localStageSet;
    }

    public StageSet getObsoleteStageSet() {
        return obsoleteStageSet;
    }

    public static String createIdentifier(Long lStageSetId, Long rStageSetId) {
        return lStageSetId + ":" + rStageSetId;
    }

    private Map<Long, Conflict> leftIdConflictsMap = new HashMap<>();
    private Map<Long, Conflict> rightIdConflictsMap = new HashMap<>();

    /**
     * will try to merge first. if it fails the merged {@link StageSet} is removed,
     * conflicts are collected and must be resolved elsewhere (eg. ask the user for help)
     *
     * @param left
     * @param right
     * @throws SqlQueriesException
     */
    @Override
    public void stuffFound(Stage left, Stage right, IFile lFile, IFile rFile) throws SqlQueriesException {
        if (!(left == null && lFile == null || left != null && lFile != null)) {
            System.err.println("ConflictSolver.stuffFound.e1");
        }
        if (!(right == null && rFile == null || right != null && rFile != null)) {
            System.err.println("ConflictSolver.stuffFound.e2");
        }

        if (left != null && right != null && (!left.getContentHash().equals(right.getContentHash()))) {
            String key = Conflict.createKey(left, right);
            if (!conflicts.containsKey(key)) {
                //todo oof for deletion
                Conflict conflict = createConflict(left, right);
                if ((left.getDeleted() ^ right.getDeleted()) || (left.getIsDirectory() ^ right.getIsDirectory())) {
                    // there might already exist a Conflict. in this case we have more detailed information now (on stage should be null)
                    // and therefore must replace it but retain the dependencies
                    if (deletedParents.get(lFile.getParentFile().getAbsolutePath()) != null) {
                        Conflict oldeConflict = deletedParents.get(lFile.getParentFile().getAbsolutePath());
                        conflict.dependOn(oldeConflict);
                    }
                    deletedParents.put(lFile.getAbsolutePath(), conflict);
                } else if (leftIdConflictsMap.containsKey(left.getParentId())) {
                    Conflict parent = leftIdConflictsMap.get(left.getParentId());
                    conflict.dependOn(parent);
                } else if (rightIdConflictsMap.containsKey(right.getParentId())) {
                    Conflict parent = leftIdConflictsMap.get(right.getParentId());
                    conflict.dependOn(parent);
                }
            }
            return;
        }
        if (left != null && right != null && left.getContentHash().equals(right.getContentHash()) && !left.getIsDirectory() && !right.getIsDirectory()) {
            Lok.debug("ConflictSolver.stuffFound.no conflict between " + left.getId() + " and " + right.getId());
        } else {
            if (left != null && leftIdConflictsMap.containsKey(left.getParentId())) {
                //conflictSearch(left, lFile, false);
                // this was set to createConflict(left, right); at some time. mybe bug related
                Conflict conflict = createConflict(left, right);
                Conflict parent = leftIdConflictsMap.get(left.getParentId());
                conflict.dependOn(parent);
            }
            if (right != null && rightIdConflictsMap.containsKey(right.getParentId())) {
                //conflictSearch(right, rFile, true);
                Conflict conflict = createConflict(left, right);
                Conflict parent = rightIdConflictsMap.get(right.getParentId());
                conflict.dependOn(parent);
            }
        }
    }

    private void conflictSearch(Stage stage, File stageFile, final boolean stageIsFromRight) {
        File directory;
        if (stage.getIsDirectory())
            directory = stageFile;
        else
            directory = stageFile.getParentFile();

        Set<String> conflictFreePaths = new HashSet<>();
        boolean resolved = false;
        while (!resolved && !directory.getAbsolutePath().equals(rootDirectory.getPath())) {
            if (deletedParents.containsKey(directory.getAbsolutePath())) {
                Conflict conflict = deletedParents.get(directory.getAbsolutePath());
                // it is proven that this file is not in conflict
                if (conflict == null) {
                    deletedParents.put(directory.getAbsolutePath(), null);
                    for (String path : conflictFreePaths)
                        deletedParents.put(path, null);
                } else {
                    Conflict dependentConflict = (stageIsFromRight ? createConflict(null, stage) : createConflict(stage, null));
                    dependentConflict.dependOn(conflict);
                    // we want to build a hierarchy of dependent conflicts
                    if (stage.getIsDirectory()) {
                        deletedParents.put(stageFile.getAbsolutePath(), dependentConflict);
                    }
                    resolved = true;
                }
            } else {
                conflictFreePaths.add(directory.getAbsolutePath());
            }
            directory = directory.getParentFile();
        }
    }

    private Conflict createConflict(Stage left, Stage right) {
        String key = Conflict.createKey(left, right);
        Conflict conflict = new Conflict(stageDao, left, right);
        conflicts.put(key, conflict);
        if (left != null)
            leftIdConflictsMap.put(left.getId(), conflict);
        if (right != null)
            rightIdConflictsMap.put(right.getId(), conflict);
        return conflict;
    }

    @SuppressWarnings("Duplicates")
    public void beforeStart(StageSet remoteStageSet) throws SqlQueriesException {
        order = new Order();
        deletedParents = new HashMap<>();
        oldeNewIdMap = new HashMap<>();
        oldeObsoleteMap = new HashMap<>();
        this.fsDao = fsDao;
        N.r(() -> {
            mergeStageSet = stageDao.createStageSet(FileSyncStrings.STAGESET_SOURCE_MERGED, remoteStageSet.getOriginCertId().v(), remoteStageSet.getOriginServiceUuid().v(), remoteStageSet.getVersion().v(), basedOnVersion);
        });
        // now lets find directories that have been deleted. so we can build nice dependencies between conflicts
        List<Stage> leftDirs = stageDao.getDeletedDirectories(lStageSetId);
        List<Stage> rightDirs = stageDao.getDeletedDirectories(rStageSetId);
        for (Stage stage : leftDirs) {
            IFile f = stageDao.getFileByStage(stage);
            Stage rStage = stageDao.getStageByPath(rStageSetId, f);
            Conflict conflict = new Conflict(stageDao, stage, rStage);
            deletedParents.put(f.getAbsolutePath(), conflict);
        }
        for (Stage stage : rightDirs) {
            IFile f = stageDao.getFileByStage(stage);
            Stage lStage = stageDao.getStageByPath(lStageSetId, f);
            Conflict conflict = new Conflict(stageDao, lStage, stage);
            deletedParents.put(f.getAbsolutePath(), conflict);
        }
    }

    private void mergeFsDirectoryWithSubStages(FsDirectory fsDirToMergeInto, Stage stageToMergeWith) throws SqlQueriesException {
        List<Stage> content = stageDao.getStageContent(stageToMergeWith.getId());
        for (Stage stage : content) {
            if (stage.getIsDirectory()) {
                if (stage.getDeleted()) {
                    fsDirToMergeInto.removeSubFsDirecoryByName(stage.getName());
                } else {
                    fsDirToMergeInto.addDummySubFsDirectory(stage.getName());
                }
            } else {
                if (stage.getDeleted()) {
                    fsDirToMergeInto.removeFsFileByName(stage.getName());
                } else {
                    fsDirToMergeInto.addDummyFsFile(stage.getName());
                }
            }
            stageDao.flagMerged(stage.getId(), true);
        }
    }

    /**
     * merges already merged {@link StageSet} with itself and predecessors (Fs->Left->Right) to find parent directories
     * which were not in the right StageSet but in the left. When a directory has changed its content
     * on the left side but not/or otherwise changed on the right
     * it is missed there and has a wrong content hash.
     *
     * @throws SqlQueriesException
     */
    public void directoryStuff() throws SqlQueriesException {
        final long oldeMergedSetId = mergeStageSet.getId().v();
        Map<Long, Long> oldeIdNewIdMapForDirectories = new HashMap<>();
        order = new Order();
        StageSet targetStageSet = stageDao.createStageSet(FileSyncStrings.STAGESET_SOURCE_MERGED, mergeStageSet.getOriginCertId().v(), mergeStageSet.getOriginServiceUuid().v(), mergeStageSet.getVersion().v(), basedOnVersion);
        N.sqlResource(stageDao.getStagesResource(oldeMergedSetId), stageSet -> {
            Stage rightStage = stageSet.getNext();
            while (rightStage != null) {
                if (rightStage.getIsDirectory()) {
                    FsDirectory contentHashDummy = fsDao.getDirectoryById(rightStage.getFsId());
                    List<Stage> content = null;

                    if (contentHashDummy == null) {
                        // it is not in fs. just add every child from the Stage
                        contentHashDummy = new FsDirectory();
                        content = stageDao.getStageContent(rightStage.getId());
                        for (Stage stage : content) {
                            if (!stage.getDeleted()) {
                                if (stage.getIsDirectory())
                                    contentHashDummy.addDummySubFsDirectory(stage.getName());
                                else
                                    contentHashDummy.addDummyFsFile(stage.getName());
                            }
                            stageDao.flagMerged(stage.getId(), true);
                        }
                    } else {
                        // fill with info from FS
                        List<GenericFSEntry> fsContent = fsDao.getContentByFsDirectory(contentHashDummy.getId().v());
                        contentHashDummy.addContent(fsContent);
                        mergeFsDirectoryWithSubStages(contentHashDummy, rightStage);
                    }
                    // apply delta
                    contentHashDummy.calcContentHash();
                    rightStage.setContentHash(contentHashDummy.getContentHash().v());
                }
                saveRightStage(rightStage, targetStageSet.getId().v(), oldeIdNewIdMapForDirectories);
                rightStage = stageSet.getNext();
            }
        });
        stageDao.deleteStageSet(oldeMergedSetId);
        stageDao.flagMergedStageSet(targetStageSet.getId().v(), false);
        mergeStageSet = targetStageSet;
    }

    private void saveRightStage(Stage stage, Long stageSetId, Map<Long, Long> oldeIdNewIdMapForDirectories) throws SqlQueriesException {
        Long oldeId = stage.getId();
        Long parentId = stage.getParentId();
        if (parentId != null && oldeIdNewIdMapForDirectories.containsKey(parentId)) {
            stage.setParentId(oldeIdNewIdMapForDirectories.get(parentId));
        }
        stage.setId(null);
        stage.setStageSet(stageSetId);
        stage.setOrder(order.ord());
        stageDao.insert(stage);
        if (stage.getIsDirectory())
            oldeIdNewIdMapForDirectories.put(oldeId, stage.getId());
    }

    public void cleanup() throws SqlQueriesException {
        //cleanup
        stageDao.deleteStageSet(rStageSetId);
        mergeStageSet.setStatus(FileSyncStrings.STAGESET_STATUS_STAGED);
        stageDao.updateStageSet(mergeStageSet);
        stageDao.deleteStageSet(obsoleteStageSet.getId().v());
        if (!stageDao.stageSetHasContent(mergeStageSet.getId().v()))
            stageDao.deleteStageSet(mergeStageSet.getId().v());
    }

    public StageSet getMergeStageSet() {
        return mergeStageSet;
    }

    /**
     * @param serverStage from server
     * @param fsStage     from fs
     * @throws SqlQueriesException
     */
    public void solve(Stage serverStage, Stage fsStage) throws SqlQueriesException {
        Stage solvedStage = null;
        final String key = Conflict.createKey(serverStage, fsStage);
        final Conflict conflict = conflicts.remove(key);
        final Long oldeRightId = fsStage == null ? null : fsStage.getId();
        final Long oldeLeftId = serverStage == null ? null : serverStage.getId();

        // if a conflict is present, apply its solution. if not, apply the left or right.
        // they must have the same hashes.
        if (conflict != null) {
            if (conflict.isLeft()) {
                solvedStage = serverStage;
                // entry exists locally, delete
                if (fsStage != null && !fsStage.getDeleted() && solvedStage == null) {
                    if (fsStage.getParentIdPair().notNull()) {
                        fsStage.setParentId(oldeObsoleteMap.get(fsStage.getParentId()));
                    }
                    // connect to the bottom fs id available in the server stageset
                    if (fsStage.getFsIdPair().isNull() && fsStage.getFsParentIdPair().isNull()) {
                        // look for oldeId first
                        Stage parent = conflict.getDependsOn().getLeft();
                        if (parent != null)
                            fsStage.setFsParentId(parent.getFsId());
                    }
                    fsStage.setDeleted(true)
                            .setId(null)
                            .setStageSet(obsoleteStageSet.getId().v())
                            .setOrder(obsoleteOrder.ord());
                    stageDao.insert(fsStage);
                    oldeObsoleteMap.put(oldeRightId, fsStage.getId());
                }
            } else {
                solvedStage = fsStage;
                // left always comes with fs ids. copy if available
                if (serverStage != null && solvedStage != null) {
                    solvedStage.setFsId(serverStage.getFsId());
                    solvedStage.setFsParentId(serverStage.getFsParentId());
                }
                // if it does not exist locally and we rejected the remote file: copy left to delete it.
                if (serverStage != null && !serverStage.getDeleted() && solvedStage == null) {
                    if (serverStage.getParentIdPair().notNull()) {
                        serverStage.setParentId(oldeNewIdMap.get(serverStage.getParentId()));
                    }
                    serverStage.setStageSet(mergeStageSet.getId().v())
                            .setId(null)
                            .setOrder(order.ord())
                            .setDeleted(true);
                    stageDao.insert(serverStage);
                }
            }
        } else if (serverStage != null) {
            solvedStage = serverStage;
        } else if (fsStage != null) {
            solvedStage = fsStage;
        } else {
            System.err.println("ConflictSolver.solve:ERRRR");
        }
        if (solvedStage != null) {
            // adapt parent stages
            // assuming that Stage.id and Stage.parentId have not been changed yet

            if (solvedStage.getParentId() != null) {
                if (oldeNewIdMap.containsKey(solvedStage.getParentId())) {
                    solvedStage.setParentId(oldeNewIdMap.get(solvedStage.getParentId()));
                } else {
                    solvedStage.setParentId(null);
                }
            }

            solvedStage.setStageSet(mergeStageSet.getId().v());
            solvedStage.setOrder(order.ord());
            solvedStage.setId(null);
            stageDao.insert(solvedStage);
            // add these things. subsequent Conflicts might reference them.
            if (oldeLeftId != null)
                oldeNewIdMap.put(oldeLeftId, solvedStage.getId());
            if (oldeRightId != null)
                oldeNewIdMap.put(oldeRightId, solvedStage.getId());
        }
    }

    private void addOldeId(Stage stage, Long newId) {

    }


    public String getIdentifier() {
        return identifier;
    }


    /**
     * check if everything is solved and sane (eg. do not modify a file when the parent folder is deleted)
     *
     * @return
     */
    public boolean isSolved() {
        Set<Conflict> isOk = new HashSet<>();
        try {
            for (Conflict conflict : conflicts.values()) {
                recurseUp(isOk, conflict);
            }
        } catch (ConflictException.UnsolvedConflictsException | ConflictException.ContradictingConflictsException e) {
            return false;
        } catch (ConflictException e) {
            System.err.println(getClass().getSimpleName() + ".isSolved().Error: " + e.getMessage());
            return false;
        }
        return true;
    }

    private void recurseUp(Set<Conflict> isOk, Conflict conflict) throws ConflictException {
        if (!conflict.hasDecision())
            throw new ConflictException.UnsolvedConflictsException();
        if (isOk.contains(conflict))
            return;
        final boolean deleted = conflict.getChoice().getDeleted();
        Stack<Conflict> stack = new Stack<>();
        stack.push(conflict);
        Conflict parent = conflict.getDependsOn();
        while (parent != null) {
            if (isOk.contains(parent))
                break;
            if (!parent.hasDecision())
                throw new ConflictException.UnsolvedConflictsException();
            if (!deleted && parent.getChoice().getDeleted())
                throw new ConflictException.ContradictingConflictsException();
            stack.push(parent);
            parent = parent.getDependsOn();
        }
        isOk.addAll(stack);
    }

    @Override
    public String toString() {
        return lStageSetId + " vs " + rStageSetId + " -> " + (mergeStageSet == null ? "null" : mergeStageSet.getId().v().toString());
    }

    public boolean hasConflicts() {
        return conflicts.size() > 0;
    }

    public Collection<Conflict> getConflicts() {
        return conflicts.values();
    }

    public void addListener(ConflictSolverListener conflictSolverListener) {
        N.r(() -> {
            listenerSemaphore.acquire();
            this.listeners.add(conflictSolverListener);
            if (obsolete)
                tellObsolete();
            listenerSemaphore.release();
        });

    }


    private void tellObsolete() {
        for (ConflictSolverListener listener : listeners)
            listener.onConflictObsolete();
    }

    /**
     * tells the {@link ConflictSolverListener}s they are obsolete if one of the merged {@link StageSet}s was part of this instance.
     *
     * @param mergedLeftStageSetId
     * @param mergedRightStageSetId
     */
    public void checkObsolete(Long mergedLeftStageSetId, Long mergedRightStageSetId) {
        if (this.lStageSetId.equals(mergedLeftStageSetId) || this.lStageSetId.equals(mergedRightStageSetId)
                || this.rStageSetId.equals(mergedLeftStageSetId) || this.rStageSetId.equals(mergedRightStageSetId)) {
            N.r(() -> {
                listenerSemaphore.acquire();
                this.obsolete = true;
                tellObsolete();
                listenerSemaphore.release();
            });
        }
    }

    public boolean isSolving() {
        return solving;
    }

    public void setSolving(boolean solving) {
        this.solving = solving;
    }

    public void setConflictHelperUuid(String conflictHelperUuid) {
        this.conflictHelperUuid = conflictHelperUuid;
    }

    public String getConflictHelperUuid() {
        return conflictHelperUuid;
    }

    public void probablyFinished() {
        if (conflictHelperUuid != null && isSolved())
            InitialIndexConflictHelper.finished(conflictHelperUuid);
    }

    public interface ConflictSolverListener {
        /**
         * called when the {@link ConflictSolver}s {@link StageSet}s were merged.
         */
        void onConflictObsolete();
    }
}

File: filesync/src/main/java/de/mel/filesync/data/conflict/ConflictSolver.kt
package de.mel.filesync.data.conflict

import de.mel.Lok
import de.mel.auth.tools.Order
import de.mel.filesync.data.FileSyncStrings
import de.mel.filesync.sql.FsDirectory
import de.mel.filesync.sql.Stage
import de.mel.filesync.sql.StageSet
import de.mel.filesync.sql.dao.ConflictDao

/**
 * Finds Conflicts according to @see bla
 * todo add link do conflict image
 */
/**
 * Basic object that finds pairs of local and remote conflicts.
 *
 * Conflicts are based on [Stage]s of a local and a remote StageSet.
 * So it may happen that only one side has a [Stage] because it does not exist in the other.
 * Conflicts can depend on another, because they are a tree structure.
 *
 * Root dependencies, conflicts without parents, are hold in [rootConflictMap].
 *
 * Keep in mind that this does not necessarily represent the actual dependency, since it is limited to [Stage]s only.
 *
 * A conflict may have multiple child conflicts on different layers that do not occur in any StageSet and therefore
 * are represented as multiple root conflicts here.
 */
open class ConflictSolver(conflictDao: ConflictDao, localStageSet: StageSet, remoteStageSet: StageSet) :
    StageSetMerger(conflictDao, localStageSet, remoteStageSet) {
    var conflictHelperUuid: String? = null
    val basedOnVersion: Long
    val conflictMap = mutableMapOf<String, Conflict>()
    val rootConflictMap = mutableMapOf<String, Conflict>()

    val localStageConflictMap = mutableMapOf<Long, Conflict>()

    val remoteStageConflictMap = mutableMapOf<Long, Conflict>()

    private val rootConflicts = mutableListOf<Conflict>()
    var order = Order()

    val idMapLocal = mutableMapOf<Long, Long>()
    val idMapRemote = mutableMapOf<Long, Long>()

    val conflictIdentifier: String

    init {
        basedOnVersion =
            if (localStageSet.basedOnVersion.v() >= remoteStageSet.basedOnVersion.v()) localStageSet.basedOnVersion.v() else remoteStageSet.basedOnVersion.v()
        conflictIdentifier = "l${localStageSet.id.v()}//r${remoteStageSet.id.v()}"
    }

    override fun before() {
        mergedStageSet = stageDao.createStageSet(
            FileSyncStrings.STAGESET_SOURCE_MERGED,
            remoteStageSet.originCertId.v(),
            remoteStageSet.originServiceUuid.v(),
            remoteStageSet.version.v() + 1,
            remoteStageSet.version.v()
        )
    }

    override fun after() {
        // calculate content hashes of all directories
        stageDao.getNotDeletedDirectoriesByStageSet(mergedStageSet.id.v()).forEach { directory ->
            val dirDummy = FsDirectory()
            val content = stageDao.getNotDeletedContent(directory.id).map { stageDao.stage2FsEntry(it).toGeneric() }
            dirDummy.addContent(content)
            val fsContent = fsDao.getContentByFsDirectory(directory.fsId)
            dirDummy.addContent(fsContent)
            dirDummy.calcContentHash()
            stageDao.updateContentHash(directory.id, dirDummy.contentHash.v())
        }
        // reset merged flag
        stageDao.flagMergedStageSet(mergedStageSet.id.v(), false)
        // make it from fs again.
        mergedStageSet.setSource(FileSyncStrings.STAGESET_SOURCE_FS);
        mergedStageSet.setStatus(FileSyncStrings.STAGESET_STATUS_STAGED)
        stageDao.updateStageSet(mergedStageSet)
        // empty stage sets must be removed
        if (!stageDao.stageSetHasContent(mergedStageSet.id.v())) {
            stageDao.deleteStageSet(mergedStageSet.id.v())
        }

    }


    fun isSolved() = conflictMap.all { it.value.hasChoice }

    private var solving: Boolean = false

    fun isSolving() = solving

    fun setSolving(v: Boolean) {
        this.solving = v
    }


    fun findConflicts(): ConflictSolver {
        /**
         * c1 conflicts have no parents since they only consist of stages that differ in contentHash and are not deleted.
         * c2 covers conflicts that have a parent-child-relation:
         *      eg. local changes file /foo/bar.txt and remote deletes /foo/
         *      this Conflict has no localStage but a parent Conflict
         *      c3 is c2 "inverted" or with the StageSets swapped.
         */
        val c1 =
            conflictDao.getC1Conflicts(localStageSetId = localStageSet.id.v(), remoteStageSetId = remoteStageSet.id.v())
        val c2 = conflictDao.getLocalDeletedByRemote(
            localStageSetId = localStageSet.id.v(),
            remoteStageSetId = remoteStageSet.id.v()
        )
        val c3 = conflictDao.getRemoteDeletedByLocal(
            localStageSetId = localStageSet.id.v(),
            remoteStageSetId = remoteStageSet.id.v()
        )
        createConflicts(c1)
        createConflicts(c2)
        createConflicts(c3)
        /**
         * find root conflicts here. they have no parent conflicts
         */
        return this
    }

    fun rec(cc: Conflict, localStage: Stage?, remoteStage: Stage?) {
        val remoteContentList: List<Stage> = if (remoteStage == null) emptyList() else stageDao.getStageContent(remoteStage.id)
        val localContentList: List<Stage> = if (localStage == null) emptyList() else stageDao.getStageContent(localStage.id)
        val remoteContent: Map<String, Stage> = remoteContentList.map { it.name to it }.toMap()
        val localContent: Map<String, Stage> = localContentList.map { it.name to it }.toMap()
        find(cc, localContent, remoteContent)
        find(cc, remoteContent, localContent, c1IsLocal = false)
    }

    fun find(con: Conflict, c1: Map<String, Stage>, c2: Map<String, Stage>, c1IsLocal: Boolean = true) {
        for ((name, c1Stage) in c1) {
            if (c2.containsKey(name)) {
                val c2Stage = c2[name]!!
                if (c2Stage.contentHash != c1Stage.contentHash) {
                    var c = if (c1IsLocal)
                        Conflict(conflictDao, c1Stage, c2Stage)
                    else
                        Conflict(conflictDao, c2Stage, c1Stage)
                    c = con.addChild(c)
                    if (c1IsLocal) {
                        localStageConflictMap[c1Stage.id] = c
                        remoteStageConflictMap[c2Stage.id] = c
                        rec(c, c1Stage, c2Stage)
                    } else {
                        localStageConflictMap[c2Stage.id] = c
                        remoteStageConflictMap[c1Stage.id] = c
                        rec(c, c2Stage, c1Stage)
                    }
                }
            } else {
                var c = if (c1IsLocal) Conflict(conflictDao, c1Stage, null) else Conflict(conflictDao, null, c1Stage)
                c = con.addChild(c)
                if (c1IsLocal) {
                    localStageConflictMap[c1Stage.id] = c
                    rec(c, c1Stage, null)
                } else {
                    remoteStageConflictMap[c1Stage.id] = c
                    rec(c, null, c1Stage)
                }
            }
        }
    }

    private fun createConflicts(dbConflicts: List<DbConflict>) {
        dbConflicts.forEach loop@{ dbConflict ->
            val localStage =
                if (dbConflict.localStageId.notNull()) stageDao.getStageById(dbConflict.localStageId.v()) else null
            val remoteStage =
                if (dbConflict.remoteStageId.notNull()) stageDao.getStageById(dbConflict.remoteStageId.v()) else null
            var conflict = Conflict(conflictDao, localStage, remoteStage)
            if (conflictMap.containsKey(conflict.key))
                return@loop
            conflictMap[conflict.key] = conflict
            rootConflictMap[conflict.key] = conflict
            if (conflict.key == "7/17")
                Lok.debug("debug tes 1")
            localStage?.let { l ->
                // Find parent conflicts
                if (l.parentIdPair.notNull() && localStageConflictMap.containsKey(l.parentId)) {
                    val parent = localStageConflictMap[l.parentId]!!
                    conflict = parent.addChild(conflict)
                    rootConflictMap.remove(conflict.key)
                }
                localStageConflictMap[l.id] = conflict
            }
            remoteStage?.let { r ->
                remoteStageConflictMap[r.id] = conflict
                // Find parent conflicts
                if (r.parentIdPair.notNull() && remoteStageConflictMap.containsKey(r.parentId)) {
                    val parent = remoteStageConflictMap[r.parentId]!!
                    conflict = parent.addChild(conflict)
                    rootConflictMap.remove(conflict.key)
                }
            }
            if (remoteStage != null && localStage != null) {
                Lok.debug("debug WIP content missing")
                if (remoteStage.isDirectory || localStage.isDirectory) {

                    rec(conflict, localStage, remoteStage)
                    Lok.debug("debug o3kaf")
                }
            }
        }
    }

    fun hasConflicts(): Boolean = !conflictMap.all { it.value.hasChoice }


    /**
     * Adds the stage to the database and creates an entry in the according idMap{Local/Remote)}.
     * The Map is determined by the StageSetId. So it has to refer to the local or remote StageSet
     *
     */
    private fun insertToMerged(decisionStage: Stage, rejectedStage: Stage?) {
        // todo debug
        if (decisionStage.namePair.equalsValue("b"))
            Lok.debug()
        val decisionStageSetId = decisionStage.stageSet
        // set parent id, but remember: parent might be from local or remote.
        decisionStage.parentId?.let { parentId ->
            // search in both maps
            (idMapLocal[parentId] ?: idMapRemote[parentId])?.let { mergedParentId ->
                decisionStage.parentId = mergedParentId
            } ?: throw ConflictException("parent id found, but parent not merged yet!")
        }
        // insert into merged StageSet
        val oldId = decisionStage.id
        decisionStage.stageSet = mergedStageSet.id.v()
        // copy new fs ids
        if (rejectedStage != null && rejectedStage.stageSet == remoteStageSet.id.v()) {
            if (rejectedStage.depth == decisionStage.depth) {
                decisionStage.fsId = rejectedStage.fsId
                decisionStage.fsParentId = rejectedStage.fsParentId
            }
            // get the fs parent id as well
            // todo performance, do something that does not query the database
            if (decisionStage.fsParentId == null) {
                stageDao.getStageById(decisionStage.parentId)?.let {
                    decisionStage.fsParentId = it.fsId
                }
            }
        }
        stageDao.insert(decisionStage)

        // insert stage first
        when (decisionStageSetId) {
            localStageSet.id.v() -> idMapLocal
            remoteStageSet.id.v() -> idMapRemote
            else -> throw ConflictException("stageset was neither local nor remote")
        }[oldId] = decisionStage.id

        // update the rejected id too
        rejectedStage?.let {
            when (rejectedStage.stageSet) {
                localStageSet.id.v() -> idMapLocal
                remoteStageSet.id.v() -> idMapRemote
                else -> throw ConflictException("stageset was neither local nor remote")
            }[rejectedStage.id] = decisionStage.id
        }
    }

    override fun foundLocal(local: Stage, remote: Stage?) {
        /**
         * check whether the conflict is a sub conflict of another one. In this case one stage set has no entry
         * and the created conflict key does not find the according conflict.
         * This occurs when one side deletes a directory while the other changes its contents.
         */
        localStageConflictMap[local.id]?.let { conflict ->
            // only apply the top
            if (conflict.decision!!.depth >= conflict.rejection!!.depth) {
                insertToMerged(conflict.decision!!, conflict.rejection)
            } else if (conflict.decision!!.deleted && conflict.decision!!.depth <= conflict.rejection!!.depth)
                return
            return
        }
        /**
         * check the remote set as well
         */
        if (remote != null)
            remoteStageConflictMap[remote.id]?.let { conflict ->
                if (conflict.decision!!.depth == conflict.rejection!!.depth) {
                    insertToMerged(conflict.decision!!, conflict.rejection)
                } else if (conflict.decision!!.deleted && conflict.decision!!.depth <= conflict.rejection!!.depth)
                    return
                return
            }
        /**
         * If conflict exists: apply its solution. Otherwise an arbitrary Stage will do, since
         * content hashes of directories have to be calculated after merging and conflicting files are in the conflict map.
         *
         */
        val key = Conflict.createKey(local, remote)
        conflictMap[key]?.let { conflict ->
            insertToMerged(conflict.decision!!, conflict.rejection)
        } ?: insertToMerged(local, remote)
    }

    override fun foundRemote(remote: Stage) {
        remoteStageConflictMap[remote.id]?.let { conflict ->
            if (conflict == null || conflict.decision == null || conflict.rejection == null)
                Lok.debug("debug fix null here")
            if (conflict.decision == conflict.remoteStage && conflict.remoteStage != null) {
                insertToMerged(conflict.decision!!, conflict.rejection)
            } else if (conflict.decision != null && conflict.rejection != null && conflict.decision!!.deleted && conflict.decision!!.depth <= conflict.rejection!!.depth)
                return
            else if (conflict.remoteStage?.deleted != true && conflict.chosenLocal) {
                // stage exists in remote but not locally. it must be flagged
                val r: Stage = conflict.remoteStage!!
                val m = Stage(this.mergedStageSet.id.v(), r)
                    .setDeleted(true)
                    .setOrder(0L)
                    .setVersion(mergedStageSet.version.v())
                    .setParentId(idMapRemote[r.parentId])
                conflictDao.stageDao.insert(m);
                Lok.debug()
            }
//            if (conflict.decision?.depth == conflict.rejection?.depth && conflict.decision!=null){
//
//            }
//            if (conflict.decision!!.depth == conflict.rejection!!.depth) {
//                insertToMerged(conflict.decision!!, null)
//            } else if (conflict.decision!!.deleted && conflict.decision!!.depth <= conflict.rejection!!.depth)
//                return
            return
        }
    }


    companion object {
        @JvmStatic
        fun createIdentifier(localStageSetId: Long, rightStageSetId: Long): String = "$localStageSetId:$rightStageSetId"
    }
}